src/compressor/brotli/BrotliCompressor.cc

```diff
@@ -0,0 +1,94 @@
+#include "brotli/encode.h"
+#include "brotli/decode.h"
+#include "BrotliCompressor.h"
+
+#define MAX_LEN (CEPH_PAGE_SIZE)
+#define MIN(a,b) (((a)<(b))?(a):(b))
+
+int BrotliCompressor::compress(const bufferlist &in, bufferlist &out) 
+{
+  unsigned have;
+  BrotliEncoderState* s = BrotliEncoderCreateInstance(NULL, NULL, NULL);
+  size_t available_in = 0;
+  const uint8_t* next_in = NULL;
+  size_t available_out = 0;
+  uint8_t* next_out = NULL;
+  if (!s) {
+      BrotliEncoderDestroyInstance(s);
+      return -1;
+  }
+  BrotliEncoderSetParameter(s, BROTLI_PARAM_QUALITY, (uint32_t)9);
+  BrotliEncoderSetParameter(s, BROTLI_PARAM_LGWIN, 22);
+  unsigned char* c_in;
+  for (std::list<buffer::ptr>::const_iterator i = in.buffers().begin(); i != in.buffers().end();) {
+    c_in = (unsigned char*) (*i).c_str();
+    size_t len = (*i).length();
+    available_in = len;
+    next_in = c_in;
+    ++i;
+    BrotliEncoderOperation finish = i != in.buffers().end() ? BROTLI_OPERATION_PROCESS : BROTLI_OPERATION_FINISH;
+    do {
+      size_t max_comp_size = BrotliEncoderMaxCompressedSize(available_in);
+      bufferptr ptr = buffer::create_page_aligned(max_comp_size);
+      next_out = (unsigned char*)ptr.c_str();
+      available_out = max_comp_size;
+      if (!BrotliEncoderCompressStream(s,finish,&available_in, &next_in, &available_out, &next_out, NULL)) {
+          BrotliEncoderDestroyInstance(s);
+          return -1;
+      }
+      have = max_comp_size - available_out;
+      out.append(ptr, 0, have);
+    } while (available_out == 0);
+    if (BrotliEncoderIsFinished(s)) {
+      BrotliEncoderDestroyInstance(s);
+      return 0;
+    }
+  }
+  BrotliEncoderDestroyInstance(s);
+  return 0;
+}
+
+int BrotliCompressor::decompress(bufferlist::iterator &p, size_t compressed_size, bufferlist &out) 
+{
+  unsigned have;
+  const char* c_in;
+  size_t available_in = 0;
+  const uint8_t* next_in = NULL;
+  size_t available_out = 0;
+  uint8_t* next_out = NULL;
+  BrotliDecoderState* s = BrotliDecoderCreateInstance(NULL, NULL, NULL);
```

Comment:
might want to use a `unique_ptr` with customized deleter or `scope_guard` for doing the cleanup.